### ISSUE TITLE
functools.partial should use PEP612 ParamSpec

### ISSUE THREAD

**Issue by benbariteau** [python/typeshed#8703]
[Currently](https://github.com/python/typeshed/blob/e2ce7c6344dc9544b3555a4b81aff5bfadc1430c/stdlib/functools.pyi#L75), functools.partial's `__call__` method takes `*args: Any, **kwargs: Any` and returns `_T` (a TypeVar for the return type of the callable the partial is wrapping). This is decent, but matching the parameter typing would make this more useful in more situations.

I can sort of see how to deal with this, although I'll admit I'm unsure how to make this generically work. I imagine you could probably deal with some number of `Concatenate`s (maybe up to 5 or 10) to deal with most cases. It's possible fixing this requires another PEP to allow an arbitrary-length `Concatenate`-type pattern.

Related to https://github.com/python/mypy/issues/1484

**Comment by AlexWaygood** [comment]
You can nearly get there with a combination of PEP 612 and PEP 646:

```python
from typing import Any, Callable, Concatenate, Generic, ParamSpec, TypeVar, TypeVarTuple, overload

_P1 = ParamSpec("_P1")
_P2 = ParamSpec("_P2")
_T = TypeVar("_T")
_R_co = TypeVar("_R_co", covariant=True)
_Ts = TypeVarTuple("_Ts")

class partial(Generic[_P1, _P2, _T, _R_co, *_Ts]):
    @overload
    def __new__(cls, __func: Callable[_P1, _R_co]) -> partial[_P1, _P1, Any, _R_co]: ...
    @overload
    def __new__(cls, __func: Callable[Concatenate[*_Ts, _P2], _R_co], *args: *_Ts) -> partial[Concatenate[*_Ts, _P2], _P2, Any, _R_co, *_Ts]: ...
    @overload
    def __new__(cls, __func: Callable[_P1, _R_co], *args: *_Ts, **kwargs: _T) -> partial[_P1, ..., _T, _R_co, *_Ts]: ...
    def __call__(self, *args: _P2.args, **kwargs: _P2.kwargs) -> _R_co: ...
    func: Callable[_P1, _R_co]
    args: tuple[*_Ts]
    keywords: dict[str, _T]
```

Mypy doesn't support PEP 646 yet, however. And even if it did, there would still be some corner cases that don't work (https://github.com/PyCQA/flake8-pyi/pull/150#issuecomment-1023575984)

**Comment by benbariteau** [comment]
Hmmm, I think that's probably the best we're gonna get for now. It's a strict improvement in my opinion.

**Comment by srittau** [comment]
Cf. #8708

**Comment by andersk** [comment]
#8708 has been closed. Can this be un-deferred yet?

**Comment by AlexWaygood** [comment]
> #8708 has been closed. Can this be un-deferred yet?

No, pytype still crashes if we make a class Generic over a TypeVarTuple, which is what the solution outlined in https://github.com/python/typeshed/issues/8703#issuecomment-1240022635 would require

**Comment by Dreamsorcerer** [comment]
> pytype still crashes

Is that the one that has apparently been fixed a couple of days ago: https://github.com/google/pytype/issues/1525#issuecomment-1847874572
Or is there another issue to track?

**Comment by AlexWaygood** [comment]
> Is that the one that has apparently been fixed a couple of days ago

no

**Comment by erikkemperman** [comment]
@AlexWaygood 
It would certainly be fantastic if the type annotations for partial captured something about the arguments as well the return type. But are you sure that the approach you outlined [above](https://github.com/python/typeshed/issues/8703#issuecomment-1240022635) would work, in principle? I didn't think it would be legal to have a `TypeVarTuple` within `Concatenate`? I'm not sure if this is just a matter of the checkers implementing it, but currently both pyright and mypy disallow this.
